modules/dhcp_common/free_ips.rb

```diff
@@ -0,0 +1,157 @@
+require 'concurrent'
+require 'logger'
+require 'set'
+
+module Proxy::DHCP
+  class FreeIps
+    include ::Proxy::Log
+
+    DEFAULT_ADDRESS_KEEP_TIME = 30*60 # 30 mins
+    DEFAULT_CLEANUP_INTERVAL = 60 # 1 min
+
+    attr_reader :allocated_ips, :allocation_timestamps, :m
+
+    def initialize
+      @allocated_ips = Set.new
+      @allocation_timestamps = []
+      @m = Monitor.new
+    end
+
+    # must be called from under the monitor
+    def mark_ip_as_allocated(ip_address)
+      @m.synchronize do
+        @allocated_ips << ip_address
+        @allocation_timestamps.push([ip_address, time_now + DEFAULT_ADDRESS_KEEP_TIME])
+      end
+    end
+
+    def clean_up_allocated_ips
+      @m.synchronize do
+        loop do
+          logger.debug("Starting allocated ip addresses cleanup pass...")
+          break if @allocation_timestamps.first.nil? || @allocation_timestamps.first[1] > time_now
+          ip, _ = @allocation_timestamps.shift
+          @allocated_ips.delete(ip)
+          logger.debug("#{ip} marked as free.")
+        end
+      end
+    end
+
+    def start
+      logger.info("Starting allocated ip address maintenance (used by unused_ip call).")
+      @timer_task = Concurrent::TimerTask.new(:execution_interval => DEFAULT_CLEANUP_INTERVAL) { clean_up_allocated_ips }
```

Comment:
Nitpick, in case of exception, will this thread (or whatever this is) survive? Shall we encapsulate this in rescue block putting down an error in logs and caring over?

Reply:
TimerTask catches all exceptions and continues/restarts the task should one occur.